Mcumgr to zcbor conversion

de-nordic opened the pull request:
Closes #40588

de-nordic committed:
mgmt/mcumgr/lib: Use zcbor in OS command group responses

The commit switches processing of OS command group responses
to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Encode STAT command group responses with zcbor

The commit moves STAT command group responses encoding to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Encode SHELL command group responses with zcbor

The commit moves SHELL command group responses encoding to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Encode FS command group responses with zcbor

The commit moves encoding of FS command group responses to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Encode IMAGE command group responses with zcbor

The commit moves encoding of IMAGE command group responses to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Add zcbor support for sending responses

The commit modifies cbor_nb_writer to handle zcbor.
Proper initialization code has been added to cbor_nb_writer_init.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Move common code for responses to zcbor

Common SMP response processing code gets switched to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
FYI @carlescufi

oyvindronningstad commented on modules/zcbor/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+
+  zephyr_include_directories(
+    ${ZEPHYR_ZCBOR_MODULE_DIR}/include
+    ${CMAKE_CURRENT_SOURCE_DIR}
```
I think this line is not needed (I don't see any header files in this directory)

de-nordic replied:
Fixed

oyvindronningstad reviewed (commented):
Quick readthrough of the zcbor stuff.

de-nordic commented on subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c:
```diff
@@ -142,17 +144,24 @@ mgmt_find_handler(uint16_t group_id, uint16_t command_id)
-		return rc;
+#ifdef CONFIG_MGMT_VERBOSE_ERR_RESPONSE
+	if (MGMT_CTXT_RC_RSN(ctxt) != NULL) {
+		rc = cbor_encode_text_stringz(&ctxt->encoder, "rsn");
```
Ah, merge fail.

de-nordic replied:
Fixed.

de-nordic commented on subsys/mgmt/mcumgr/buf.c:
```diff
@@ -144,7 +146,14 @@ cbor_nb_write(struct cbor_encoder_writer *writer, const char *data, int len)
+	cnw->zs.payload_mut = nb->data + sizeof(struct mgmt_hdr);
+	cnw->zs.payload_end = nb->data + net_buf_tailroom(nb);
+	cnw->zs.elem_count = 0;
+	cnw->zs.indefinite_length_array = false;
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -7,14 +7,27 @@
+		zcbor_tstr_put_lit(zse, label) && zcbor_bool_put(zse, value)
+#else
+/* In "frugal" lists flags are added to response only when they evaluate to true */
+/* Note that value is evaluated tiwce! */
```
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -7,14 +7,27 @@
 
+#ifndef CONFIG_IMG_MGMT_FRUGAL_LIST
+#define ZCBOR_ENCODE_FLAG(zse, label, value)					\
+		zcbor_tstr_put_lit(zse, label) && zcbor_bool_put(zse, value)
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/stat_mgmt/src/stat_mgmt.c:
```diff
@@ -87,39 +138,33 @@ stat_mgmt_show(struct mgmt_ctxt *ctxt)
+	ok = zcbor_tstr_put_lit(zse, "rc")		&&
+	     zcbor_int32_put(zse, MGMT_ERR_EOK)		&&
+	     zcbor_tstr_put_lit(zse, "stat_list")	&&
+	     zcbor_map_start_encode(zse, 50);
```
It did not make difference for mcumgr-cli so got unnoticed. Fixed now.

oyvindronningstad commented on subsys/mgmt/mcumgr/lib/smp/src/smp.c:
```diff
@@ -55,57 +60,34 @@ smp_read_hdr(struct smp_streamer *streamer, struct mgmt_hdr *dst_hdr)
+	struct cbor_nb_writer *nbw = (struct cbor_nb_writer *)streamer->mgmt_stmr.writer;
+	zcbor_state_t *zsp = &nbw->zs;
+
+	zcbor_map_start_encode(zsp, 1);
+	zcbor_tstr_put_lit(zsp, "rc");
+	zcbor_int32_put(zsp, status);
```
Ok, makes sense, I'd recommend either a comment or an assert statement to this effect. BTW zcbor recently gained an error variable in the state struct.

lairdjm replied:
Would add in the check, it's possible that an application has a memory corruption issue and if this is triggered, it can be apparent at the time that something is very wrong

lairdjm commented on subsys/mgmt/mcumgr/lib/smp/src/smp.c:
```diff
@@ -160,20 +132,21 @@ smp_handle_single_payload(struct mgmt_ctxt *cbuf, const struct mgmt_hdr *req_hdr
 
 	if (handler_fn) {
 		*handler_found = true;
+		zcbor_map_start_encode(&cbuf->cnbe->zs, 10);
```
Would make this a Kconfig

de-nordic replied:
Gone to Kconfgi.

de-nordic committed:
mgmt/mcumgr/lib: Remove cbor_encoder_writer from usage

With transition to zcbor, the cbor_encoder_writer structure of
TinyCBOR is no longer used.
This commit replaces the structure with cbor_nb_writer, which
gathers zcbor processing structure with netbuf into one object.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Replace cbor_decoder_reader with cbor_nb_reader

Replaces TinyCBOR cbor_decoder_reader with cbor_nb_reader in
SMP processing code.
The SMP source code has been relying on internal elements of the
cbor_decoder_reader and had to be rewritten to use net_buf
structures, that are part of cbor_nb_reader, instead.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Replace calls to mgmt_streamer_init_writer

The calls have been replaced with direct invocations of
cbor_nb_writer_init.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Switch cbor_nb_reader to zcbor

The commit removes cbor_decoder_reader from cbo_nb_reader and
adds zcbor state variable to the structure.
All the code that has been supporting the cbor_decder_reader
has been removed and/or replaced with zcbor/net_buf specific.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Switch smp_process_request_packet to net_buf

Drops usage of mgmt_streamer.reader in favour of direct access
to net_buf.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Remove TinyCBOR state objects from mgmt_ctxt

Removes no longer needed elements of struct mgmt_ctxt.
Removes mgmt_ctxt_init, mgmt_streamer_init_reader and init_reader
from struct mgmt_streamer_cfg as they are no longer needed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Switch OS group request processing to zcbor

The commit switches OS group commands to use zcbor instead
of TinyCBOR for processing SMP requests.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Move STAT group request processing to zcbor

The STAT command group request processing has been moved from
TinyCBOR to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Move SHELL mgmt request processing to zcbor

The commit replaces TinyCBOR for request processing within
SHELL command group.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Add utility zcbor_map_decode_bulk

The commit adds private, for internal mcumgr use only, utility
function for bulk processing of CBOR map entries.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Move IMG mgmt group request processing to zcbor

The TinyCBOR, in processing SMP requests, have been replaced
with zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -26,6 +29,21 @@ static struct {
+{
+	bool ok;
+
+	ok = zcbor_tstr_put_lit(zse, "rc")	&&
+	     zcbor_int32_put(zse, rc)		&&
+	     zcbor_tstr_put_lit(zse, "off")	&&
+	     zcbor_uint64_put(zse, off);
```
Nope. Encoding via CDDL wouldn't make sense, the code would get larger. The decoding has been planned to be moved to CDDL, but as the CDDL right now does not support decoding of payload where position of fields changes it is postponed.
The significant improvement though is that zcbor uses zcbor_string to point where data is placed in decoded buffer, for example the mcumgr buffer, which means that it is no longer needed to copy for example chunks of file image update to additional buffer, mcumgr transport buffer is used directly. This saves on stack an static variables, and copy operations.
Some Kconfig options will disappear.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -178,90 +191,61 @@ img_mgmt_state_read(struct mgmt_ctxt *ctxt)
+		ok = zcbor_tstr_put_term(zse, vers_str)					&&
+		     zcbor_tstr_put_term(zse, "hash")					&&
+		     zcbor_bstr_encode(zse, &zhash)					&&
+		     ZCBOR_ENCODE_FLAG(zse, "bootable", flags & IMAGE_F_NON_BOOTABLE)	&&
```
Logic is fine. zcbor functions return true on success.

lairdjm replied:
From the original code: `err |= cbor_encode_boolean(&image, !(flags & IMAGE_F_NON_BOOTABLE));`
The flag is inverted which is missing here

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -121,109 +127,98 @@ os_mgmt_taskstat_encode_stack_info(struct CborEncoder *thread_map,
-
-	return err;
+	return os_mgmt_taskstat_encode_thread_name(zse, idx, thread)	&&
+	       zcbor_map_start_encode(zse, 10)				&&
+	       os_mgmt_taskstat_encode_priority(zse, thread)		&&
+	       zcbor_tstr_put_lit(zse, "tid")				&&
+	       zcbor_uint32_put(zse, idx)				&&
+	       zcbor_tstr_put_lit(zse, "state")				&&
+	       zcbor_uint32_put(zse, thread->base.thread_state)		&&
+	       os_mgmt_taskstat_encode_stack_info(zse, thread)		&&
+	       os_mgmt_taskstat_encode_unsupported(zse)			&&
+	       zcbor_map_end_encode(zse, 10);
```
Right now yes, but I will increase it.

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -121,109 +127,98 @@ os_mgmt_taskstat_encode_stack_info(struct CborEncoder *thread_map,
-		return MGMT_ERR_ENOMEM;
-	}
+	zcbor_tstr_put_lit(zse, "tasks");
+	zcbor_map_start_encode(zse, 50);
```
Will make this configurable.

de-nordic committed:
mgmt/mcumgr/lib: Move FS command group request processing to zcbor

The commit moves processing of SMP requests from TinyCBOR to zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Move Zephyr specific group to zcbor

Replacing TinyCBOR with zcbor.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Remove mgmt_err_from_cbor function

No longer needed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Remove cborattr from mcumgr

No longer used by mcumgr.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Remove dependency on TinyCBOR

Mcumgr now wholy uses zcbor for decoding/encoding SMP
requests/responses.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Deprecate FS_MGMT_UL_CHUNK_SIZE

Now the chunk of uploaded file is directly read from the transport
buffer, net_buf, and is no longer copied into intermediate buffer
so the real size of chunk is actually how much the sender could
fit into MCUMGR_BUF_SIZE, less the other fields.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Deprecate CONFIG_OS_MGMT_ECHO_LENGTH

The echo sentence is now directly copied from request transport
buffer to response buffer, without use of intermediate buffer.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Deprecate CONFIG_IMG_MGMT_UL_CHUNK_SIZE

The IMG_MGMT_UL_CHUNK_SIZE is no longer used to control size of
intermediate buffer used for application image chunk processing,
instead directly request trasport buffer is used.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -12,57 +12,72 @@
+	}
+
+	zcbor_tstr_put_lit(&ctxt->cnbe->zs, "r");
+	if (!zcbor_tstr_encode(&ctxt->cnbe->zs, &value)) {
```
If a map is provided that doesn't contain the correct echo request data, what happens at this point?

de-nordic replied:
The zcbor_str_decode line that decodes echo request into value would fail if it would not be able to match payload to tstr; so we should never reach this line if payload is not right.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -121,109 +136,97 @@ os_mgmt_taskstat_encode_stack_info(struct CborEncoder *thread_map,
 	}
 #endif /* CONFIG_INIT_STACKS */
 #endif /* CONFIG_THREAD_STACK_INFO */
+	ok = zcbor_tstr_put_lit(zse, "stksiz")		&&
+	     zcbor_int64_put(zse, stack_size)		&&
+	     zcbor_tstr_put_lit(zse, "stkuse")		&&
+	     zcbor_int64_put(zse, stack_unused);
```
Can/should these be negative?

de-nordic replied:
Will fix.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/stat_mgmt/src/stat_mgmt.c:
```diff
@@ -95,45 +115,69 @@ stat_mgmt_cb_encode(CborEncoder *enc, struct stat_mgmt_entry *entry)
 
-	err |= cbor_encode_text_stringz(&ctxt->encoder, "fields");
-	err |= cbor_encoder_create_map(&ctxt->encoder, &map_enc, CborIndefiniteLength);
+	memcpy(stat_name, value.value, value.len);
```
As above, value isn't sure to be set if it wasn't included in the cbor map and will have undefined data

de-nordic replied:
Fixed

lairdjm replied:
The way I see this is: `struct zcbor_string value;` value is undefined, `while()` loop processes all items in message, assuming that the message does contain valid data but no "name" tag and instead is "timber", `ok` will be true but `value` will never be set and will be undefined, so at this point `value.len` and `value.value` could be anything

de-nordic replied:
Fixed.

lairdjm commented on include/mgmt/mcumgr/buf.h:
```diff
@@ -8,18 +8,18 @@
 struct cbor_nb_reader {
-	struct cbor_decoder_reader r;
 	struct net_buf *nb;
+	zcbor_state_t zs[4];
```
Significance of 4?

de-nordic replied:
Number of zcbor state backups, will add Kconfig for this.

de-nordic replied:
Added Kconfigs for that

carlescufi commented:
@mniestroj @nandojve a review from you would be most welcome.

de-nordic commented on subsys/mgmt/mcumgr/buf.c:
```diff
@@ -25,130 +28,24 @@ mcumgr_buf_free(struct net_buf *nb)
 	cnw->nb->len = sizeof(struct mgmt_hdr);
-	cnw->enc.bytes_written = sizeof(struct mgmt_hdr);
-	cnw->enc.write = &cbor_nb_write;
+	zcbor_new_encode_state(&cnw->zs, 1, nb->data + sizeof(struct mgmt_hdr),
```
Fixed

de-nordic commented on modules/zcbor/Kconfig:
```diff
@@ -0,0 +1,10 @@
+config ZEPHYR_ZCBOR_MODULE
+	bool
+
+config ZCBOR
```
Please log the comment here https://github.com/zephyrproject-rtos/zephyr/pull/43332

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/stat_mgmt/src/stat_mgmt.c:
```diff
@@ -61,24 +111,25 @@ stat_mgmt_show(struct mgmt_ctxt *ctxt)
+	     zcbor_tstr_put_lit(zse, "name")		&&
+	     zcbor_tstr_put_term(zse, stat_name)	&&
+	     zcbor_tstr_put_lit(zse, "fields")		&&
+	     zcbor_map_start_encode(zse, 50);
```
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:
```diff
@@ -280,32 +268,30 @@ img_mgmt_erase(struct mgmt_ctxt *ctxt)
+	 * to check return codes from zcbor as failing to encode so short
+	 * response means that CONFIG_MCUMGR_BUF_SIZE is too short to even
+	 * handle requests.
+	 */
```
Added checks, removed comment.

carlescufi commented:
@lairdjm could you please take another look?

de-nordic commented:
@lairdjm @oyvindronningstad @nandojve @mniestroj please review.

lairdjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -26,6 +26,84 @@ config APP_LINK_WITH_MCUMGR
+	  A group or command that adds additional maps/lists above the
+	  base map, which is already taken into account, should
+	  select one of the MGMT_MIN_DECODING_LEVEL_?.
+	default 0 if !MGMT_MIN_DECODING_LEVEL_1	&&	\
+		     !MGMT_MIN_DECODING_LEVEL_2 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_3 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_4 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_5
+	default 1 if MGMT_MIN_DECODING_LEVEL_1	&&	\
+		     !MGMT_MIN_DECODING_LEVEL_2 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_3 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_4 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_5
+	default 2 if MGMT_MIN_DECODING_LEVEL_2 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_3 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_4 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_5
+	default 3 if MGMT_MIN_DECODING_LEVEL_3 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_4 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_5
+	default 4 if MGMT_MIN_DECODING_LEVEL_4 &&	\
+		     !MGMT_MIN_DECODING_LEVEL_5
+	default 5 if MGMT_MIN_DECODING_LEVEL_5
```
Would this be tidier/easier to maintain with new levels if it was created the other way around?
```
 	default 5 if MGMT_MIN_DECODING_LEVEL_5
 	default 4 if MGMT_MIN_DECODING_LEVEL_4
 	default 3 if MGMT_MIN_DECODING_LEVEL_3
 	default 2 if MGMT_MIN_DECODING_LEVEL_2
 	default 1 if MGMT_MIN_DECODING_LEVEL_1
 	default 0
```

de-nordic replied:
Fixed

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -54,33 +59,31 @@ fs_mgmt_file_download(struct mgmt_ctxt *ctxt)
 	char path[CONFIG_FS_MGMT_PATH_SIZE + 1];
-	unsigned long long off;
-	CborError err;
+	unsigned long long off = ULLONG_MAX;
```
can uint64_t be used now that tinycbor has done?

de-nordic replied:
Fixed.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -12,57 +12,70 @@
-	err = cbor_encode_text_stringz(&ctxt->encoder, "r")				||
-	      cbor_encode_text_stringz(&ctxt->encoder, echo_buf);
+	ok = zcbor_tstr_put_lit(zse, "r")		&&
+	     zcbor_tstr_encode(zse, &value);
```
This is where I see a potential issue, if there has been some sensitive data pushed on the stack and the underlying data in `value` happens to have valid-ish data, some/all of that sensitive data might be returned to this command

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:
```diff
@@ -73,15 +86,16 @@ os_mgmt_taskstat_encode_thread_name(struct CborEncoder *encoder, int idx,
 	}
 
-	return cbor_encode_text_string(encoder, thread->name, name_len);
+	return zcbor_tstr_encode(zse,
+			&(struct zcbor_string){.value = thread->name,
+					       .len = name_len });
```
OK, will use zcbor_tstr_encode_ptr .

de-nordic replied:
Fixed

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -26,6 +26,70 @@ config APP_LINK_WITH_MCUMGR
+	default 2 if MGMT_MIN_DECODING_LEVEL_2
+	default 3 if MGMT_MIN_DECODING_LEVEL_3
+	default 4 if MGMT_MIN_DECODING_LEVEL_4
+	default 5 if MGMT_MIN_DECODING_LEVEL_5
```
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/util/include/zcbor_bulk/zcbor_bulk_priv.h:
```diff
@@ -0,0 +1,115 @@
+extern "C" {
+#endif
+
+struct zcbor_map_decode_key {
+	struct zcbor_string key;	/* Map key string */
+	zcbor_decoder_t *decoder;	/* Key corresponding decoder */
+};
+
+struct zcbor_map_decode_val {
+	void *value_ptr;
+	bool found;
+};
```
@oyvindronningstad  Stats for current solution as give here (2022-04-08, 0655+00), with pair of constant-variable maps:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      201064 B       412 KB     47.66%
            SRAM:       46934 B       256 KB     17.90%
        IDT_LIST:          0 GB         2 KB      0.00%
```
build after proposed changes with single map created as auto-variable:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      201088 B       412 KB     47.66%
            SRAM:       46934 B       256 KB     17.90%
        IDT_LIST:          0 GB         2 KB      0.00%
```
Build for nrf52840dk_nrf52840 with all mcumgr groups enabled. The only change is visible in flash and does not seem to be significant enough to keep complexity of two separate maps for preforming translation.

carlescufi commented:
@lairdjm could you take another look?

de-nordic commented:
@lairdjm Sorry for dragging you here again; I had to fix the zcbor_map_decode_bulk where, do to merge error, I have been returning "false" instead of -EBADMSG.

lairdjm commented on subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:
```diff
@@ -120,50 +112,36 @@ fs_mgmt_file_download(struct mgmt_ctxt *ctxt)
-	off = ULLONG_MAX;
-	rc = cbor_read_object(&ctxt->it, uload_attr);
-	if (rc != 0 || off == ULLONG_MAX || file_name[0] == '\0') {
+	ok = zcbor_map_decode_bulk(zsd, fs_upload_decode,
+		ARRAY_SIZE(fs_upload_decode), &decoded) == 0;
```
`ok` is a bool and `zcbor_map_decode_bulk` is returning an int, is a case needed?

de-nordic replied:
no ==  does the cast

lairdjm replied:
Missed that!
